### ISSUE TITLE
Deterministic color selection

### ISSUE THREAD

**Issue by apinstein** [metabase/metabase#5910]
We do a lot of charts and dashboards. We have noticed that the color assignments are non-deterministic w/r/t the value represented. So for instance, a chart that includes a "foo" value might be blue on one chart, and on another chart which also uses "foo" as a value, it might be green.

I suggest that you use a deterministic approach to selecting color assignments based on the represented value, which will yield the following benefits:

 I feel this feature would drastically reduce the cognitive load of interpreting charts, especially when using dashboards.

- dashboards that have multiple widgets will have consistency of how colors are represented by default across widgets
- a chart's colors won't change when sort order is changed
- a chart's colors will be consistent even when values are missing (ie foo=green regardless of whether or not "bar" and "baz" appear in the chart)

The latter seems a bit harder, since you'll have to map colors independently (vs just indexing into a palette). I would certainly settle for nailing the first 2 by hashing values, sorting them, and using those as an index into the color lookup.

Please let me know if I can clarify this.

**Comment by salsakran** [comment]
this seems reasonable. Thoughts @mazameli  + @kdoh ?

**Comment by mazameli** [comment]
This seems pretty crisp to me, though I think there are some interesting details hiding in the corners:
- What constitutes a "value," and how do we decide two values are the same? E.g., if one chart is count of users by OS, and another chart is average time on site by OS, those are two different metrics broken out by the same dimension — so would the respective `macOS` series be the same color or a different color on those two charts?
- If we evaluate sameness by the metric/aggregation, does the presence/absence of filters on two metrics affect things? E.g., if I have `count of orders by day where state is CA` and `count of orders by day where state is FL`, I'm assuming those would have distinct colors; but what about `count of orders by day, last 30 days` vs. `count of orders by day, last 60 days`?
- If one chart on the dashboard has count of orders by product over time, and another chart is a pie chart of count of orders by product, I would expect the broken out series and pie segments to match.
- We currently let users change the color of lines/areas/bars (but not pie slices). If I change the color of a series in one chart, I would expect that it would probably change every instance of that series/metric on the dashboard.
- If I add a new chart to an existing dashboard, and the chart includes a metric/aggregation that's already present on other cards on the dashboard, does my new chart change color to match?

**Comment by MarkRx** [comment]
It looks like it happens because the database stores the visualization settings as a list of colors. The colors appear to be applied in the order that new values are encountered from the results. You can order the results to keep the ordering consistent but if you have a subset of all possible results then colors get shifted.

For example suppose I have 3 statuses that are always in the following order:

- Failure Red (#EF8C8C)
- Success Green (#9cc177)
- Warning Yellow (509EE3)

The settings get persisted in the database as followed:
```
{
	"graph.dimensions": ["JB_AGENT", "JB_STATUS"],
	"graph.metrics": ["count"],
	"graph.colors": ["#EF8C8C", "#9cc177", "#509EE3", . . . (other colors)]
}
```

Now suppose I add some filter that results in there being no failures that show up. Colors are assigned as new values as found so I end up with:

1. Success - Red (#EF8C8C)
2. Warning - Green (#9cc177)

Yuck! I am now showing Success as red and warning as green??

In order to function properly the way the data is saved needs to change such that colors remember their assigned value.

**Comment by HelenMertes** [comment]
It is also unpractical that a new coulour is automaticall assigned, if there is an additional series added to for example a bar chart a new coulour is automaticall assigned, but this new colour is then not necessarily available to choose from if I want to change it manually. 
Hard work (or impossible) then to align colours in different questions for the same category to make a dashboard user-/reader friendly.
Thanks for your work on this!

**Comment by adjenks** [comment]
This issue, and issue #6143, are so painful to deal with, that we will not use Metabase for any real work until the issue is resolved. I was really looking forward to using it, but when you assign colours to groups, like blue to ice cream and red to hotdog, I expect blue to represent ice cream all the time, not to pseudo-randomly change depending on which value appears in the data first. My managers will be angry with me if this happens. I am not complaining at all, just stressing how important this seemingly small issue is. I would love make Metabase one of our primary analytics tools, so I am really looking forward to hotdog data not being coloured blue ever again.

**Comment by mfurmaniuk** [comment]
I will just add my $0.03 to this in that I have the same issues with selecting colors to appear as standard.  I use trend reports that can adjust depending on Test Groups, for a good many of them they appear correct with Pass as Green and Fail as Red, we have some Skips but not every group has those.  I've had managers tell me they get some tables appear where the colors are switched, where Pass gets Red and Fail gets Green.  On a visual cue that's just wrong.

I've seen on different monitors, with the same question, the colors get switched.  Plus at times where a test group completely fails I don't  need to see Green show up, as that will confuse users.

It would be nice if, when using Filter Drop Downs (and coming in on the resulting links) that the colors stay standard on the report regardless of how the values come back.

I like Metabase, and we do a lot with it, but things like this while seemingly small have a big impact when dealing with Managers and above.

**Comment by MarkRx** [comment]
This is on my list of things to do. I plan on contributing back a fix. It would probably convert graph.colors from an array to an array of objects like this:

```
{
color: ...
value: ...
type: (number, boolean, etc)
}
```

**Comment by m3l1k4** [comment]
hi @MarkRx was there ever a fix for this? I tried googling it but it seems to not pick up results on it

**Comment by flamber** [comment]
@m3l1k4 Please use the forum for questions and troubleshooting: https://discourse.metabase.com/

**Comment by adjenks** [comment]
@m3l1k4 If you look just above your post, you can see that a branch was merged into the master code branch on Sep 17, 2018. If you click on it you can see that it says "Resolves #5910 Deterministic color selection".  So I'm pretty sure that it was fixed in any version released after that date. If you have a version from before that date, then you need to update. If it still looks broken after installing the latest version, then you should open a ticket and provide examples of how to reproduce the error so that it can be fixed.
